FIAT/gopalakrishnan_lederer_schoberl.py

```diff
@@ -0,0 +1,154 @@
+
+        sd = ref_el.get_spatial_dimension()
+        if size is None:
+            size = sd
+
+        shape = (size, size)
+        num_exp_functions = expansion_set.get_num_members(degree)
+        num_components = size * size - 1
+        num_members = num_components * num_exp_functions
+        embedded_degree = degree
+
+        # set up coefficients for traceless tensors
+        basis = traceless_matrices(ref_el)
+        coeffs_shape = (num_members, *shape, num_exp_functions)
+        coeffs = numpy.zeros(coeffs_shape, "d")
+        cur_bf = 0
+        for S in basis.reshape(-1, *shape):
+            for exp_bf in range(num_exp_functions):
+                coeffs[cur_bf, :, :, exp_bf] = S
+                cur_bf += 1
+
+        super().__init__(ref_el, degree, embedded_degree, expansion_set, coeffs)
+
+
+def GLSSpace(ref_el, degree):
+    """Return the subspace of trace-free Pk tensors with normal-tangential
+    component in P_{k-1}"""
+    sd = ref_el.get_spatial_dimension()
+    P = TracelessTensorPolynomialSet(ref_el, degree, variant="bubble")
+    expansion_set = P.get_expansion_set()
+    if degree == 1:
+        dimP1 = expansion_set.get_num_members(degree)
+        coeffs = numpy.zeros((sd+1, sd-1, 2, sd+1, sd-1, dimP1))
+        for i, j in numpy.ndindex(coeffs.shape[0:2]):
+            # Constant times traceless matrix
+            coeffs[i, j, 0, i, j, :] = 1
+            # Barycentric coordinate times traceless matrix
+            coeffs[i, j, 1, i, j, i] = 1
+        coeffs = coeffs.reshape(-1, P.get_num_members())
+        coeffs = numpy.tensordot(coeffs, P.get_coeffs(), axes=(1, 0))
+        return polynomial_set.PolynomialSet(ref_el, degree, degree, expansion_set, coeffs)
+    else:
+        entity_ids = polynomial_entity_ids(ref_el, degree, continuity=expansion_set.continuity)
+        mask = numpy.ones((P.get_num_members(),), int).reshape(sd+1, sd-1, -1)
+        constrained_ids = {}
+        for dim in entity_ids:
+            constrained_ids[dim] = {}
+            if dim == 0 or dim == sd:
+                for entity in entity_ids[dim]:
+                    constrained_ids[dim][entity] = []
+            else:
+                dimPkm1 = len(ref_el.make_points(dim, 0, degree-1))
+                for entity in entity_ids[dim]:
+                    constrained_ids[dim][entity] = entity_ids[dim][entity][dimPkm1:]
+
+        closure_ids = dual_set.make_entity_closure_ids(ref_el, constrained_ids)
+        for facet in closure_ids[sd-1]:
+            mask[facet, ..., closure_ids[sd-1][facet]] = 0
+        indices = numpy.flatnonzero(mask)
+        return P.take(indices)
+
+
+class GLSDual(dual_set.DualSet):
+    def __init__(self, ref_el, degree):
+        sd = ref_el.get_spatial_dimension()
+        top = ref_el.get_topology()
+        nodes = []
+        entity_ids = {dim: {entity: [] for entity in sorted(top[dim])} for dim in sorted(top)}
+
+        # Face dofs: bidirectional nt Legendre moments
```

Comment:
Is this general dimension (1 normal + (sd-1) tangents)?  If so, the `nt` comment is misleading.  If not and this is just 2, the code is over engineered?

Reply:
It's for general dimension, I changed the comment to something more accurate: "moments of nt components against a basis for P_{k-1}"

Reply:
But if you still say `nt components` that's a bit ambiguous.  I'd recommend something more like `moments of the normal component and tangential component(s) against a basis for P_{k-1} on the facet`